cassiopeia/plotting/utilities.py

```diff
@@ -0,0 +1,525 @@
+    the returned dictionaries have (thetas, radii) as its elements, which are
+    the angles and radii in polar coordinates respectively.
+
+    Note:
+        This function only *places* (i.e. computes the x and y coordinates) a
+        tree on a coordinate system and does no real plotting.
+
+    Args:
+        tree: The CassiopeiaTree to place on the coordinate grid.
+        depth_key: The node attribute to use as the depth of the nodes. If
+            not provided, the distances from the root is used by calling
+            `tree.get_distances`.
+        orient: The orientation of the tree. Valid arguments are `left`, `right`,
+            `up`, `down` to display a rectangular plot (indicating the direction
+            of going from root -> leaves) or any number, in which case the
+            tree is placed in polar coordinates with the provided number used
+            as an angle offset.
+        depth_scale: Scale the depth of the tree by this amount. This option
+            has no effect in polar coordinates.
+        width_scale: Scale the width of the tree by this amount. This option
+            has no effect in polar coordinates.
+        extend_branches: Extend branch lengths such that the distance from the
+            root to every node is the same. If `depth_key` is also provided, then
+            only the leaf branches are extended to the deepest leaf.
+        angled_branches: Display branches as angled, instead of as just a
+            line from the parent to a child.
+        polar_interpolation_threshold: When displaying in polar coordinates,
+            many plotting frameworks (such as Plotly) just draws a straight line
+            from one point to another, instead of scaling the radius appropriately.
+            This effect is most noticeable when two points are connected that
+            have a large angle between them. When the angle between two connected
+            points in polar coordinates exceed this amount (in degrees), this function
+            adds additional points that smoothly interpolate between the two
+            endpoints.
+        polar_interpolation_step: Interpolation step. See above.
+        add_root: Add a root node so that only one branch connects to the
+            start of the tree. This node will have the name `synthetic_root`.
+
+    Returns:
+        Two dictionaries, where the first contains the node coordinates and
+        the second contains the branch coordinates.
+    """
+    root = tree.root
+    nodes = tree.nodes.copy()
+    edges = tree.edges.copy()
+    depths = None
+    if depth_key:
+        depths = {
+            node: tree.get_attribute(node, depth_key) for node in tree.nodes
+        }
+    else:
+        depths = tree.get_distances(root)
+
+    placement_depths = {}
+    positions = {}
+    leaf_i = 0
+    leaves = set()
+    for node in tree.depth_first_traverse_nodes(postorder=False):
+        if tree.is_leaf(node):
```

Comment:
I'm using DFS here on purpose, because I need a specific ordering of leaves, such that siblings are grouped together. This is necessary for plotting because otherwise there may be branches that overlap!

Reply:
I have another question -- how difficult would it be to "reorder" leaves such that some criteria is met? Like if you have a huge rake of leaves that have the same character state and you want to group together leaves that have the same cluster ID? 

It almost seems like it would be as easy as just grouping first by the attribute in the cell_meta variable and then outputting coordinates of leaves by category.

Reply:
Do you mean reordering the leaves within a clade according to some variable? I think that would require quite a bit of bookkeeping to make sure that no branches overlap in the visualization.